resources/templates/md/posts/2016-04-04-state.md

```diff
@@ -0,0 +1,370 @@
+and subsequently calling the appropriate functions to create new states.
+
+NOTE:
+
+If you don't have any knowledge about threads in Clojure, you should read about [threads](http://blog.clojuregirl.com/clojure_threads.html) first.
+
+## How we can manage states (or how the data might change over time)
+<div>
+<div id="marshmallow-text">
+
+There are four ways how to model an identity (manage states).
+Using reference types: `Vars` , `Atoms` , `Refs` , `Agents` .
+
+Greg has a girlfriend Gregorina and they go to the party. It's halloween party and
+there are a lot of sweets everywhere. Llamas love candies and as we know, 
+favourite candies for llamas are yummy marshmallows and gummy bears.
+</div>
+<div id="marshmallow-picture">
+![marshmallow](../img/marshmallow.png "Marshmallow")
+</div>
+</div>
+### Vars
+
+We can create a `Var` using `def`.
+In our example we create a `Var` named `marshmallows-in-the-bag` `(def marshmallows-in-the-bag)`
+and this `Var` will be reference for amount of the marshmallows in one bag.
+
+If we define `Var` as a `^:dynamic` (adding the `:^dynamic` metadata part).
+
+```clojure
+(def ^:dynamic *marshmallows-in-the-bag* 30)
+```
+
+`Var` can be rebound in different context and the rebinding is available only within it's dynamic scope.
+In Clojure, dynamic vars are typically surrounded with `*` in what is called the earmuff convention.
+
+We will look what the dynamic scope is, and on the difference between Lexical and Dynamic scope.
+
+### Difference between the lexical and the dynamic scope
+
+What's the scope 
+
+Simply, the `scope` is a context created to summarize symbols. Without this scope all the symbols would be global.
+
+Let's start with the lexical scope
+
+### Lexical Scope
+
+Is the code block, where we define the symbol. This symbol is valid inside the code block where is defined (including any nested forms).
+The lexical scope can be defined for example by the `let` block.
+
+In the bag there are 30 marshmallows. So we will define `marshmallows-in-the-bag` with a value `30`.
+
+```clojure
+(def marshmallows-in-the-bag 30)
+```
+
+Now we will define a function `marshmallows-left`. This function takes one argument `marshmallows-eaten`. 
+In a let block we will define the `new-amount-of-marshmallows` and the value of this Var is the difference between `marshmallows-in-the-bag`
+and `marshmallows-eaten`. 
+As a last step in this function is the println `new-amount-of-marshmallows`.
+
+```clojure
+(defn marshmallows-left [marshmallows-eaten]
+  (let [new-amount-of-marshmallows (- marshmallows-in-the-bag marshmallows-eaten)]
+     (println (str "Marshmallows in the bag: " new-amount-of-marshmallows))))
+```
+
+Calling our function `marshmallows-left` with 10 (because Greg ate 10 pieces of marshmallow)
+
+```Clojure
+(marshmallows-left 10)
+
+;;=> Marshmallows in the bag: 20
+```
+We found out that in the bag 20 marshmallows left.
```

Comment:
Here too, I would put this sentence above, as a continuation of the first, otherwise it's difficult to read.
"If we call `marshmallows-left` with the argument `10` (because Greg ate 10 pieces of marshmallow), we find out that there are 20 marshmallows left in the bag."